### ISSUE TITLE
Particle emitter overflow causes entity rendering to fail

### ISSUE THREAD

**Issue by thegrb93** [Facepunch/garrysmod-issues#1968]
Once a particle emitter overflows, for some reason all entities will no longer render except for the map and you need to reload the map in order to fix it.

There is console spam with
```
CUtlLinkedList overflow! (exhausted index range) (65535)
```
Which I think originates from here https://github.com/ValveSoftware/source-sdk-2013/blob/55ed12f8d1eb6887d348be03aee5573d44177ffb/mp/src/game/client/particlemgr.cpp#L1302 

But the way garrysmod handles the error isn't working.

**Comment by thegrb93** [comment]
I've found that citadels can also trigger this effect.

```
lua_run for I=1,200 do e=ents.Create("prop_physics") e:SetModel("models/props_combine/combine_citadel001b.mdl") e:SetPos(Entity(1):GetEyeTrace().HitPos+Vector(0,0,3000)) e:Spawn() end
```

**Comment by thegrb93** [comment]
After removing them, things go back to normal at least.

**Comment by X-Coder** [comment]
I have the same issue on my servers since the release of garrysmod 13, didn't have the issue before garrysmod 13. It happens on my windows and linux dedicated server.

Whenever people are using guns which spawns bullets (bullets using the ParticleEmitter in LUA) every prop spawned in the map stops rendering and goes invisible. The invisible props will still have collision. The map stays visible. But moving map parts, like tracks or doors become invisible too.
Cleanup the map doesn't fix the issue, only map reload helps.

I have seen this issue for example on spacebuild weapons, and lots of the stargate weapons of carter's addon pack. Basically every weapon which makes use of the ParticleEmitter for creating bullets.

Here is one implementation of ParticleEmitter triggering this bug:
https://github.com/X-Coder/SBEP-Weapons/blob/master/lua/effects/tinypulsesplode/init.lua

Can you please check how the ParticleEmitter cleanup is currently working in garrysmod/engine?
Maybe the Finish-function is not cleaning up / freeing up the memory (in time)?

**Comment by X-Coder** [comment]
Is it related to this?
https://facepunch.com/showthread.php?t=1309609#2
> The bug is due to Garry's Mod not removing the previous emitter. Calling Finish will not remove the particle emitter. The particle emitter appears to be left dangling in memory thus causing the overflow.

Is there a universal fix for the base lua code?

**Comment by robotboy655** [comment]
Are you not calling Emitter:Finish() in your code?

**Comment by thegrb93** [comment]
The generic_smoke effect is the main culprit for my server I think

**Comment by robotboy655** [comment]
It's not a Lua effect, it's an Orange Box .pcf effect.

**Comment by thegrb93** [comment]
I know

**Comment by X-Coder** [comment]
robotboy655, I linked a lua code above which triggers the bug. 
Emitter:Finish() is used there already in line 45, but it makes no difference.

**Comment by thegrb93** [comment]
Revisiting this issue. https://github.com/garrynewman/garrysmod/blob/master/garrysmod/gamemodes/sandbox/entities/entities/gmod_emitter.lua#L125 Is it because there's no Emitter:Finish() for this ParticleEffect?

**Comment by robotboy655** [comment]
That's an orange box particle, it has nothing to do with Lua Particles

**Comment by thegrb93** [comment]
According to https://facepunch.com/showthread.php?t=1350845, you need to emitter:Finish()

**Comment by Bo98** [comment]
ParticleEffect ≠ ParticleEmitter

**Comment by thegrb93** [comment]
Ah sorry, bad reading.

**Comment by thegrb93** [comment]
A friend of mine made a video showing how smoke emitters can affect the number of active particle systems. He uses cl_show_num_particle_systems 1 to show the number of active particle systems. It shows that the number of particle systems can continue increasing to infinity, causing broken rendering and crashes.

https://www.youtube.com/watch?v=mPq2nNn28xU&feature=youtu.be

**Comment by X-Coder** [comment]
Thank you for making this video. This is exactly what I have seen too on my and other servers. So the rendering bug happens when there are ~850 active particle systems as you can see in the video at 2:40.

Is there a convar to limit active particle systems? Or can this be fixed lua side?

**Comment by Python1320** [comment]
Has anyone figured which linked list it even is that is overflowing? 
Two guesses:
 - https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/sp/src/game/client/clientleafsystem.cpp#L299
 - https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/game/client/particlemgr.h#L746

Maybe it is possible to recover from it.

**Comment by robotboy655** [comment]
Pretty sure it's the renderables one. The more props/entities are on the map, the easier it is do.

**Comment by X-Coder** [comment]
Don't know but can the integer data type of this CUtlLinkedList be changed to something else?

**Comment by Kefta** [comment]
@X-Coder It's not an integer limit. It's an unsigned short

**Comment by X-Coder** [comment]
@Kefta Thank you, on closer look I have seen it now too.

There is this comment for the [CUtlLinkedList ](https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/public/tier1/utllinkedlist.h#L53)class:
```C++
// Class S is the storage type; the type you can use to save off indices in 
// persistent memory. Class I is the iterator type, which is what should be used
// in local scopes. I defaults to be S, but be aware that on the 360, 16-bit 
// arithmetic is catastrophically slow. Therefore you should try to save shorts
// in memory, but always operate on 32's or 64's in local scope.
// The ideal parameter order would be TSMI (you are more likely to override M than I)
// but since M depends on I we can't have the defaults in that order, alas.
template <class T, class S = unsigned short, bool ML = false, class I = S, class M = CUtlMemory< UtlLinkedListElem_t<T, S>, I > > 
class CUtlLinkedList{...}
```
The slowness warning in the comment is only about the 360, which doesn't matter for garrysmod anyway.

Could you try to change the iterator type to unsigned int or long to get around the exhausted index range 65535 error in the lines Python1320 posted above? This would at least delay it, before it errors out again.

**Comment by thegrb93** [comment]
I imagine you can't just change the default index type of a container used widely throughout source engine without consequences. The root of the problem needs to be fixed, otherwise it'll just cause the game to run out of memory rather than overflow the container.

**Comment by X-Coder** [comment]
You are right, the root of the problem is what it needs to be fixed, or its gonna overflow sooner or later again. Increasing the data type also increases memory usage, which is not really what we want.

But somewhere we need a starting point. This error shows up almost daily on my server since the release of garrysmod 13 in 2013, before that I can't remember to show up.

If you look through the sources, the lists are used only in some simple for each loops. There is only one for loop that is using int data type. 

As Python1320 said, we need first to find out which one it is that is overflowing. How we can do this? Can you maybe add some debug code, which get printed together with the "CUtlLinkedList overflow!" error message so we can locate the one causing this error?

Here is another use of CUtlLinkedList which has to do with the particle's materials:
https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/game/client/particlemgr.h#L593

 My hope is you can find a way to fix it. With all the brains here there has to be a way :)

**Comment by robotboy655** [comment]
So the problem with the "Generic Smoke" effect is that it creates a new particle system for each particle. This is obviously wrong, I think I will add support for continuous OB particle systems for the Emitter and make a continuous smoke particle system, that should fix this exploit. Doesn't fix the under laying issue, but it can't really be properly fixed, there will always be a limit of concurrent renderables.

**Comment by thegrb93** [comment]
Yeah we know it creates a new particle system each time, but why it doesn't free the particle system once the smoke disappears (after there are too many) is a mystery.

**Comment by robotboy655** [comment]
It does, but you can spam enough of them before they despawn.

**Comment by X-Coder** [comment]
Why didn't it break in this way before garrysmod 13? -at least I can't remember any map breaks like this one. Is it related to a garrysmod 13 update or a change in Valve's engine code?

If you have access to old garrysmod sources you could check how garrysmod used the particle emitter before 13?

**Comment by Kefta** [comment]
It did -- maybe not from particle emitters but I remember random linked list errors; it was just way less common. No idea why

**Comment by robotboy655** [comment]
Because there were no "Generic Smoke" effect that could spam in insane amounts of particle systems/emitters.

**Comment by besenyei** [comment]
I used the code from @thegrb93, spawned the citadel.

> CUtlLinkedList overflow! (exhausted index range) (65535)
> CShadowMgr::AddShadowDecalToSurface - overflowed m_ShadowDecals linked list!

This error then happens, because there's too many shadows being rendered to the surface because of that citadel.
